daemon/api_sideload_n_try.go

```diff
@@ -270,6 +270,57 @@ func sideloadManySnaps(st *state.State, snapFiles []*uploadedSnap, flags sideloa
 	return chg, nil
 }
 
+var readComponentInfoFromCont = readComponentInfoFromContImpl
+
+func readComponentInfoFromContImpl(tempPath string) (*snap.ComponentInfo, error) {
+	compf, err := snapfile.Open(tempPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open container: %w", err)
+	}
+	return snap.ReadComponentInfoFromContainer(compf)
+}
+
+// readComponentInfo reads ComponentInfo from a snap component file and the
+// snap.Info of the matching installed snap. If instanceName is not empty, it
+// is used to find the right instance, otherwise the SnapName from the
+// component is used.
+func readComponentInfo(st *state.State, tempPath, instanceName string, flags sideloadFlags) (*snap.ComponentInfo, *snap.Info, *apiError) {
+	if !flags.dangerousOK {
+		// TODO read assertions for components
+		return nil, nil, BadRequest("only local components can be installed at the moment")
+	}
+
+	ci, err := readComponentInfoFromCont(tempPath)
+	if err != nil {
+		return nil, nil, BadRequest("cannot read component metadata: %v", err)
+	}
+
+	// If no instance was provided in the request we use the snap name from the component
+	if instanceName == "" {
+		instanceName = ci.Component.SnapName
+	}
+	si, err := installedSnapInfo(st, instanceName)
+	if err != nil {
+		return nil, nil, BadRequest("snap for component not installed: %v", err)
```

Comment:
shouldn't we detect the right errors that mean the snap is missing vs other errors? also shouldn't we use the apiError equivalent of snap.NotInstalledError ?

Reply:
I've changed this a bit, also because I want to distinguish between when the error is because the file is not really a component from other errors.